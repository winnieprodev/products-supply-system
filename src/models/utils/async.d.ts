export interface Error {
  message: string;
}

export interface Async<T> {
  isLoading: boolean;
  error: Error | undefined;
  data: T | undefined;
}
