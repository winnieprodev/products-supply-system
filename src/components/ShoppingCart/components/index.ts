export * from "./CartAnchor";
export * from "./CartDropdown";
