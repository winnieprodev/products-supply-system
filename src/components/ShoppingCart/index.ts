export { default as ShoppingCart } from "./ShoppingCart";
export { ShoppingCartContext } from "./utils";
export { default as ShoppingCartWrapper } from "./ShoppingCartWrapper";
