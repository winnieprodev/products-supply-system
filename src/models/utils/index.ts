export * from "./async.d";
export * from "./enums";
