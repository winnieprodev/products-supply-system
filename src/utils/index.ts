export * from "./routeUtils";
export * from "./supplierUtils";
export * from "./testUtils";
