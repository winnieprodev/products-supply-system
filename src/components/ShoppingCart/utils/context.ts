import { Async, Cart, ProductCartInfo } from "models";
import React from "react";
import { mockAsync, mockCart } from "utils";

interface ShoppingCartContextProps {
  cartId: Async<number>;
  cart: Async<Cart> | undefined;
  setCart: (cart: Async<Cart> | undefined) => void;
  setProduct: (product: ProductCartInfo) => void;
  removeProduct: (productId: string) => void;
}

export const EMPTY_CART = { products: [] };
const mockFunction = () => {};

export const ShoppingCartContext = React.createContext<ShoppingCartContextProps>(
  {
    cartId: mockAsync(0),
    cart: mockAsync(mockCart(0)),
    setCart: mockFunction,
    setProduct: mockFunction,
    removeProduct: mockFunction,
  },
);
