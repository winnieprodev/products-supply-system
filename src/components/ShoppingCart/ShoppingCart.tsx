import { Dropdown } from "components";
import React, { useContext, useEffect } from "react";
import { Empty } from "utils/types";

import { ShoppingCartContext } from "./utils";
import { CartAnchor, CartDropdown } from "./components";
import { useGetCart } from "queries";
import { AsyncWrapper } from "components/AsyncWrapper";
import { WrapperMode } from "models";

interface ShoppingCartContent {
  cartId: number;
}

const ShoppingCartContent: React.FC<ShoppingCartContent> = ({ cartId }) => {
  const cart = useGetCart(cartId);

  const { setCart } = useContext(ShoppingCartContext);

  useEffect(() => {
    setCart(cart);
    return () => setCart(undefined);
  }, [cart]);

  return (
    <AsyncWrapper item={cart} mode={WrapperMode.tiny}>
      {cart.data && (
        <Dropdown
          closeOnRouteSwitch
          anchor={<CartAnchor amount={cart.data.length} />}
        >
          <CartDropdown cart={cart.data} />
        </Dropdown>
      )}
    </AsyncWrapper>
  );
};

const ShoppingCart: React.FC<Empty> = () => {
  const { cartId } = useContext(ShoppingCartContext);
  if (cartId.data === undefined) {
    return null;
  }
  return <ShoppingCartContent cartId={cartId.data} />;
};

export default ShoppingCart;
