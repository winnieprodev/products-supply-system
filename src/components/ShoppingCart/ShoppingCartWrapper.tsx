import { Async, Cart, ProductCartInfo } from "models";
import React, { useCallback, useState } from "react";
import { Empty } from "utils/types";
import { useSetCartItem } from "queries";

import { ShoppingCartContext } from "./utils";
import { useCartId } from "./hooks/useCartId";

export const ShoppingCartWrapper: React.FC<React.PropsWithChildren<Empty>> = ({
  children,
}) => {
  const [setCartItem, removeCartItem] = useSetCartItem();

  const cartId = useCartId();

  const [cart, setCart] = useState<Async<Cart> | undefined>();

  const setProduct = useCallback(
    (product: ProductCartInfo) => {
      cartId.data && setCartItem(cartId.data, product);
    },
    [cartId.data],
  );

  const removeProduct = useCallback(
    (productId: string) => {
      cartId.data && removeCartItem(cartId.data, productId);
    },
    [cartId.data],
  );
  return (
    <ShoppingCartContext.Provider
      value={{ cart, setCart, cartId, setProduct, removeProduct }}
    >
      {children}
    </ShoppingCartContext.Provider>
  );
};

export default ShoppingCartWrapper;
