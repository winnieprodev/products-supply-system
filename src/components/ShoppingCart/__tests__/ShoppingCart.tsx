import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { mockAsync } from "utils";

import ShoppingCart from "../ShoppingCart";
import ShoppingCartWrapper from "../ShoppingCartWrapper";

jest.mock("@apollo/client", () => ({
  gql: (req: string) => req,
  useMutation: () => [jest.fn(), mockAsync(undefined, true)],
}));

describe("ShoppingCart", () => {
  describe("Snapshot tests", () => {
    test("it matches the snapshot", () => {
      const { container } = render(
        <BrowserRouter>
          <ShoppingCartWrapper>
            <ShoppingCart />
          </ShoppingCartWrapper>
        </BrowserRouter>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
