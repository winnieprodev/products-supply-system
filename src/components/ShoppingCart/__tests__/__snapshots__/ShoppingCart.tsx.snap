// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart Snapshot tests it matches the snapshot 1`] = `null`;
